File: src/reply-receipt.ts
import { TextDecoder } from 'node:util';
import type { RequestEndpoint } from './request-service.js';
import { hasLoneSurrogate } from './domain/text-content.js';

export const MAX_REPLY_RECEIPT_LENGTH = 8192;
const MAX_RECEIPT_STRING_BYTES = 4096;

export interface ReplyReceipt {
  readonly version: 1;
  readonly requestId: string;
  readonly attemptId: string;
  readonly endpoint: RequestEndpoint;
}

export type ReplyReceiptErrorCode = 'RESPONSE_RECEIPT_INVALID' | 'RESPONSE_RECEIPT_MISMATCH';

export class ReplyReceiptError extends Error {
  constructor(
    public readonly code: ReplyReceiptErrorCode,
    message: string,
    options?: { cause?: unknown }
  ) {
    super(message, options);
    this.name = 'ReplyReceiptError';
  }
}

function invalid(message = 'Response receipt is invalid.', cause?: unknown): ReplyReceiptError {
  return new ReplyReceiptError('RESPONSE_RECEIPT_INVALID', message, { cause });
}

function assertBoundedString(value: unknown, label: string): asserts value is string {
  if (
    typeof value !== 'string' ||
    value.length === 0 ||
    hasLoneSurrogate(value) ||
    Buffer.byteLength(value, 'utf8') > MAX_RECEIPT_STRING_BYTES
  ) {
    throw invalid(`Response receipt ${label} is invalid.`);
  }
}

function assertEndpoint(endpoint: unknown): asserts endpoint is RequestEndpoint {
  if (!endpoint || typeof endpoint !== 'object' || Array.isArray(endpoint)) {
    throw invalid('Response receipt endpoint is invalid.');
  }
  const value = endpoint as Record<string, unknown>;
  const keys = Object.keys(value).sort();
  if (
    keys.join('\u0000') !==
    ['paneId', 'panePid', 'serverId', 'serverPid', 'serverStartTime', 'socketPath'].join('\u0000')
  ) {
    throw invalid('Response receipt endpoint is invalid.');
  }
  assertBoundedString(value.serverId, 'serverId');
  assertBoundedString(value.socketPath, 'socketPath');
  assertBoundedString(value.serverStartTime, 'serverStartTime');
  assertBoundedString(value.paneId, 'paneId');
  if (!/^%\d+$/.test(value.paneId)) throw invalid('Response receipt paneId is invalid.');
  if (!Number.isSafeInteger(value.serverPid) || (value.serverPid as number) <= 0) {
    throw invalid('Response receipt serverPid is invalid.');
  }
  if (!Number.isSafeInteger(value.panePid) || (value.panePid as number) <= 0) {
    throw invalid('Response receipt panePid is invalid.');
  }
}

function validateReceipt(value: unknown): ReplyReceipt {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw invalid();
  }
  const record = value as Record<string, unknown>;
  const keys = Object.keys(record).sort();
  if (keys.join('\u0000') !== ['attemptId', 'endpoint', 'requestId', 'version'].join('\u0000')) {
    throw invalid();
  }
  if (record.version !== 1) throw invalid('Response receipt version is invalid.');
  assertBoundedString(record.requestId, 'requestId');
  if (Buffer.byteLength(record.requestId, 'utf8') > 256) {
    throw invalid('Response receipt requestId is invalid.');
  }
  assertBoundedString(record.attemptId, 'attemptId');
  assertEndpoint(record.endpoint);
  return {
    version: 1,
    requestId: record.requestId,
    attemptId: record.attemptId,
    endpoint: {
      serverId: record.endpoint.serverId,
      socketPath: record.endpoint.socketPath,
      serverPid: record.endpoint.serverPid,
      serverStartTime: record.endpoint.serverStartTime,
      paneId: record.endpoint.paneId,
      panePid: record.endpoint.panePid,
    },
  };
}

/** Encode a versioned, canonical, unpadded base64url correlation receipt. */
export function encodeReplyReceipt(value: ReplyReceipt): string {
  const receipt = validateReceipt(value);
  const encoded = Buffer.from(JSON.stringify(receipt), 'utf8').toString('base64url');
  if (encoded.length > MAX_REPLY_RECEIPT_LENGTH) {
    throw invalid('Response receipt exceeds the maximum length.');
  }
  return encoded;
}

/** Decode and validate a receipt; positional request ID correlation is checked when supplied. */
export function decodeReplyReceipt(encoded: string, requestId?: string): ReplyReceipt {
  if (
    typeof encoded !== 'string' ||
    encoded.length === 0 ||
    encoded.length > MAX_REPLY_RECEIPT_LENGTH ||
    !/^[A-Za-z0-9_-]+$/.test(encoded)
  ) {
    throw invalid();
  }
  let decoded: string;
  try {
    decoded = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }).decode(
      Buffer.from(encoded, 'base64url')
    );
  } catch (error) {
    throw invalid('Response receipt is not valid UTF-8.', error);
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(decoded);
  } catch (error) {
    throw invalid('Response receipt is not valid JSON.', error);
  }
  const receipt = validateReceipt(parsed);
  if (Buffer.from(decoded, 'utf8').toString('base64url') !== encoded) {
    throw invalid('Response receipt is not canonical.');
  }
  if (requestId !== undefined && receipt.requestId !== requestId) {
    throw new ReplyReceiptError(
      'RESPONSE_RECEIPT_MISMATCH',
      'Response receipt does not match the requested ID.'
    );
  }
  return receipt;
}

export function validateReplyRequestId(value: unknown): asserts value is string {
  if (typeof value !== 'string' || value.length === 0 || hasLoneSurrogate(value)) {
    throw new Error('Request ID must be a non-empty string.');
  }
  if (Buffer.byteLength(value, 'utf8') > 256) {
    throw new Error('Request ID must not exceed 256 UTF-8 bytes.');
  }
}

File: rust/crates/tmt-adapters/src/reply_receipt.rs
//! Bounded wire decoding only. Lookup and proof validation belong to the
//! request service transaction; this adapter never consults storage or tmux.

use base64::{Engine, engine::general_purpose::URL_SAFE_NO_PAD};
use serde_json::{Map, Value};
use std::{error::Error, fmt};
use tmt_core::{
    endpoint::ServerEvidence,
    limits::MAX_JS_SAFE_INTEGER,
    request::{RequestEndpoint, ResponseProof, correlation::response_token},
};

pub const MAX_REPLY_RECEIPT_LENGTH: usize = 8192;
pub const COMPACT_REPLY_RECEIPT_LENGTH: usize = 25;
const COMPACT_PREFIX: &str = "v2_";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ReplyReceiptError {
    Invalid,
    Mismatch,
}

impl ReplyReceiptError {
    pub fn code(self) -> &'static str {
        match self {
            Self::Invalid => "RESPONSE_RECEIPT_INVALID",
            Self::Mismatch => "RESPONSE_RECEIPT_MISMATCH",
        }
    }
}

impl fmt::Display for ReplyReceiptError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Self::Invalid => "Response receipt is invalid.",
            Self::Mismatch => "Response receipt does not match the requested ID.",
        })
    }
}
impl Error for ReplyReceiptError {}

/// Call with an immutable prepared association, never the current live pane.
pub fn encode_short_receipt(
    request_id: &str,
    attempt_id: &str,
    endpoint: &RequestEndpoint,
) -> String {
    format!(
        "{COMPACT_PREFIX}{}",
        URL_SAFE_NO_PAD.encode(response_token(request_id, attempt_id, endpoint))
    )
}

pub fn decode_reply_receipt(
    encoded: &str,
    request_id: &str,
) -> Result<ResponseProof, ReplyReceiptError> {
    if encoded.is_empty() || encoded.len() > MAX_REPLY_RECEIPT_LENGTH {
        return Err(ReplyReceiptError::Invalid);
    }
    if let Some(payload) = encoded.strip_prefix(COMPACT_PREFIX) {
        if encoded.len() != COMPACT_REPLY_RECEIPT_LENGTH {
            return Err(ReplyReceiptError::Invalid);
        }
        let bytes = URL_SAFE_NO_PAD
            .decode(payload)
            .map_err(|_| ReplyReceiptError::Invalid)?;
        return Ok(ResponseProof::Compact(
            bytes.try_into().map_err(|_| ReplyReceiptError::Invalid)?,
        ));
    }
    // v1 is decode-only. Canonical BASE64 spelling does not require canonical
    // JSON order or whitespace: the previous decoder accepted both.
    let bytes = URL_SAFE_NO_PAD
        .decode(encoded)
        .map_err(|_| ReplyReceiptError::Invalid)?;
    let text = std::str::from_utf8(&bytes).map_err(|_| ReplyReceiptError::Invalid)?;
    let value: Value = serde_json::from_str(text).map_err(|_| ReplyReceiptError::Invalid)?;
    decode_v1(&value, request_id)
}

fn decode_v1(value: &Value, request_id: &str) -> Result<ResponseProof, ReplyReceiptError> {
    let record = object(value, &["version", "requestId", "attemptId", "endpoint"])?;
    if record["version"].as_f64() != Some(1.0) {
        return Err(ReplyReceiptError::Invalid);
    }
    let recorded_request = bounded_string(&record["requestId"], 256)?;
    let attempt_id = bounded_string(&record["attemptId"], 4096)?;
    let endpoint = object(
        &record["endpoint"],
        &[
            "serverId",
            "socketPath",
            "serverPid",
            "serverStartTime",
            "paneId",
            "panePid",
        ],
    )?;
    let pane_id = bounded_string(&endpoint["paneId"], 4096)?;
    if !pane_id
        .strip_prefix('%')
        .is_some_and(|digits| !digits.is_empty() && digits.bytes().all(|b| b.is_ascii_digit()))
    {
        return Err(ReplyReceiptError::Invalid);
    }
    let endpoint = RequestEndpoint {
        server: ServerEvidence {
            server_id: bounded_string(&endpoint["serverId"], 4096)?.into(),
            socket_path: bounded_string(&endpoint["socketPath"], 4096)?.into(),
            server_pid: pid(&endpoint["serverPid"])?,
            server_start_time: bounded_string(&endpoint["serverStartTime"], 4096)?.into(),
        },
        pane_id: pane_id.into(),
        pane_pid: pid(&endpoint["panePid"])?,
    };
    if recorded_request != request_id {
        return Err(ReplyReceiptError::Mismatch);
    }
    Ok(ResponseProof::Recorded {
        attempt_id: attempt_id.into(),
        endpoint,
    })
}

fn object<'a>(
    value: &'a Value,
    keys: &[&str],
) -> Result<&'a Map<String, Value>, ReplyReceiptError> {
    value
        .as_object()
        .filter(|map| map.len() == keys.len() && keys.iter().all(|key| map.contains_key(*key)))
        .ok_or(ReplyReceiptError::Invalid)
}

fn bounded_string(value: &Value, max: usize) -> Result<&str, ReplyReceiptError> {
    value
        .as_str()
        .filter(|text| !text.is_empty() && text.len() <= max)
        .ok_or(ReplyReceiptError::Invalid)
}

fn pid(value: &Value) -> Result<u64, ReplyReceiptError> {
    // JSON.parse used IEEE-754 numbers, including integer-valued 1.0 / 1e0.
    let number = value.as_f64().ok_or(ReplyReceiptError::Invalid)?;
    if !number.is_finite()
        || number < 1.0
        || number > MAX_JS_SAFE_INTEGER as f64
        || number.fract() != 0.0
    {
        return Err(ReplyReceiptError::Invalid);
    }
    Ok(number as u64)
}

#[cfg(test)]
mod tests;
